### ISSUE TITLE
Stopped Working on macOS Safari 13

### ISSUE THREAD

**Issue by razor-1** [google/google-api-javascript-client#589]
All of the sudden, with no changes on my side, Google sign in is broken on Safari 13.0.4 running on macOS 10.15.2.

I started getting reports from users in recent days and am able to reproduce it myself. The popup appears with the consent form/account chooser, you proceed, see a spinner for a second that says "one moment please" then the popup disappears, and none of the callbacks fire. There are no errors logged to the console that I can see when this happens.  It works fine on Safari on iOS, macOS Chrome, etc.

Did something change recently?

**Comment by matt-mcdaniel** [comment]
I'm experiencing the same and wondering if it's due to Safari updates and new cookie policies:
https://webkit.org/blog/9521/intelligent-tracking-prevention-2-3/. However, I can't get their ITP Debug tool to work :/.

**Comment by razor-1** [comment]
For me even if I interact with accounts.google.com, e.g. by signing out, and then go back through the flow, it's still broken. In the past with ITP issues, interacting with the domain prevented the issue.
Despite this being the official client, I have to believe that high volume sites are using something else so they aren't randomly stuck with users who can't log in.

**Comment by matt-mcdaniel** [comment]
@razor-1 are you running your code on a CMS or platform? I'm using Sharepoint and wondering if that's contributing to the issue.

**Comment by razor-1** [comment]
No. This is a site I control top to bottom.

**Comment by danielx** [comment]
Anyone managed to find a workaround yet without migrating to the server side sign in flow?
We have multiple apps that use the client side sign in flow with this library.

**Comment by eesheesh** [comment]
Can you provide steps to reproduce the issue from scratch, i.e. on a fresh webpage?

The sample in https://developers.google.com/identity/sign-in/web/sign-in doesn't work for me in Safari, but if I create a simple webpage based on the guide, it does work in Safari:
```html
<script src="https://apis.google.com/js/platform.js" async defer></script>
<meta name="google-signin-client_id" content="REPLACEME.apps.googleusercontent.com">
<div class="g-signin2" data-onsuccess="onSignIn"></div>
<script>
function onSignIn(googleUser) {
  var profile = googleUser.getBasicProfile();
  console.log('ID: ' + profile.getId()); // Do not send to your backend! Use an ID token instead.
  console.log('Name: ' + profile.getName());
  console.log('Image URL: ' + profile.getImageUrl());
  console.log('Email: ' + profile.getEmail()); // This is null if the 'email' scope is not present.
}
</script>
```

**Comment by danielx** [comment]
@eesheesh which version of Mac OS X are you running?

Our customers started noticing this issue after upgrading OS X to v10.15.2 from v10.15.1.
I'll create a small reproducible sample tomorrow when i get to the office.

**Comment by eesheesh** [comment]
I'm on macOS 10.15.2 with Safari 13.0.4. I'll wait for your sample, thanks!

**Comment by matt-mcdaniel** [comment]
I was able to fix the issue by making sure that Preferences->Privacy "Prevent cross-site tracking" was _**checked**_. Never thought to try this because it seemed so counterintuitive, quite the Safari 🐛 .

**Comment by danielx** [comment]
@matt-mcdaniel Thanks. Though it doesn't seem to make it work in Private Window browsing even with "Prevent cross-site tracking" enabled.

@eesheesh as matt said the bug also seems to be dependent on having "Prevent cross-site tracking" disabled in Safari privacy preferences. Though it seems that if you try to sign in to a website using this javascript client library in a Private Window it doesn't work regardless of that setting.

Our main issue with this bug is that we use [Safe Exam Browser](https://github.com/SafeExamBrowser/seb-mac) for locking down student devices in exam environments and Safe Exam Browser uses Safari to display websites. Sadly the "Prevent cross-site tracking" Safari preference doesn't fix the issue in Safe Exam Browser.

**Comment by razor-1** [comment]
I can confirm that *enabling* "Prevent cross-site tracking" fixes this. Is anyone who owns this repo looking at fixing it, or should we just abandon this client?

**Comment by eesheesh** [comment]
I tried this on my [sample above](https://github.com/google/google-api-javascript-client/issues/589#issuecomment-572639154), and it works regardless of whether "Prevent cross-site tracking" is enabled or not (I deleted cookies for my webserver's domain between tests).

If someone can add an example page that can reproduce the issue, I'll try this on my system as well.

**Comment by matt-mcdaniel** [comment]
@danielx yes a private window fails for me regardless. 

@eesheesh you are on 13.0.4 right? Can you try unchecking "prevent cross-site tracking", clearing data, quitting safari and reopening and trying again? Multiple people on my team have been able to reproduce the issue.

@razor-1 this seems to be an issue with Safari rather than this repo, I know that there are open tickets with Apple for this issue. Doesn't seem to be specific to google apis.

**Comment by danielx** [comment]
@matt-mcdaniel do you have any links to these open tickets?
I looked at the webkit issue tracker but i couldn't find a ticket that i thought was created for this issue since i'm not sure what the root cause is.

**Comment by eesheesh** [comment]
Thanks @matt-mcdaniel - I can now reproduce this after following your steps (using the same test webpage I've used before).

I looked around a bit more and it seems like there are some known issues with this feature of Safari on 13.0.4, for example: https://discussions.apple.com/thread/250942525.

Based on that, I think it's best to wait for Apple to issue a fix. I can't reproduce the issue on [Safari Technology Preview](https://developer.apple.com/safari/download/) Release 98 (Safari 13.1, WebKit 15609.1.13.4), so I assume this is going to be fixed in Safari soon.

**Comment by wittyapps** [comment]
I too am having the same issues with Safari.  To confirm it wasn't my code, I copied the sample auth html page and replaced with my credentials with no success.  I will disable Google features on my web application for Safari browsers until it is resolved.

google-api-javascript-client/samples/authSample.html

**Comment by brada1703** [comment]
I can confirm that this bug exists in Safari 13.0.3. 😥
However, it doesn't exist in Safari 11.1.1 🎉

**Comment by brada1703** [comment]
Temporary Fix:

`var using_safari = navigator.userAgent.indexOf("Safari") != -1;
if(using_safari) {
    alert('There is a bug in Safari 13.* Please try with Google Chrome')
}`

🤣🤣

**Comment by scottie-schneider** [comment]
Same here, Safari V. 13.4.5.
Means I have to change the entire way I authenticate. Thanks apple.

**Comment by yuzhakovvv** [comment]
If you have "access_token" for example after using another Google Sign in method you can use [Google API CORS](https://github.com/google/google-api-javascript-client/blob/master/docs/cors.md#making-authenticated-requests) to access Google API authenticated requests

```
function asyncXhr(method, url) {
  return new Promise(function(resolve, reject) {
    const xhr = new XMLHttpRequest();

    xhr.open(method, url);

    xhr.onload = function() {
      if (this.status >= 200 && this.status < 300) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };

    xhr.onerror = function() {
      reject(JSON.parse(xhr.response));
    };

    xhr.send();
  });
}

await asyncXhr(
  'GET',
  `https://www.googleapis.com/gmail/v1/users/${email}/threads/${threadId}?access_token=${encodeURIComponent(accessToken)}`
);
```

**Comment by roxigonzalez** [comment]
> Same here, Safari V. 13.4.5.
> Means I have to change the entire way I authenticate. Thanks apple.

Unfortunately I have to do the same :(

**Comment by eesheesh** [comment]
@yuzhakovvv, please note that adding `access_token` to your query string will be [deprecated in 2021](https://developers.google.com/identity/protocols/oauth2#4.-send-the-access-token-to-an-api.). I'd suggest using `setRequestHeader()` on the XHR and add the token to the `Authorization` header, like this:

```
xhr.setRequestHeader('Authorization', 'Bearer ' + accessToken);
```

**Comment by thientran1707** [comment]
a workaround that I find out is to use `//apis.google.com/js/platform.js` instead of the lightweight SDK `//apis.google.com/js/api.js`
I have to do a check for Safari browser and load this script instead

**Comment by tombrowndev** [comment]
Similar issue on Chrome IOS.

When 'Allow Cross-Site Tracking' setting is turned off, the sign-in page shows 'One moment please', immediately returns and does not send the callback. 

Sign in does work when 'Allow Cross-Site Tracking' is turned on.

@thientran1707 I was also using `api.js`and have tried switching to `platform.js` but no luck.

I cannot reproduce the issue in Mac OS Chrome & Safari or iOS Safari. Another strange observation is that it does work for initial login (after clearing browsing data).

**Comment by tombrowndev** [comment]
Here is a sample of what I am doing (simplified a bit for ease of understandinig)

```javascript
// On page load

window.gapi.load("client:auth2", async () => {
       await window.gapi.client.init({
            clientId: "XXXXXXXXX",
            scope: "email",
       });

       auth = window.gapi.auth2.getAuthInstance();
 });

// On sign in click

if (auth.current.isSignedIn.get()) {
        await auth.current.signOut(); // I force sign in at the moment.
}

await auth.signIn();
const tokenId = auth.currentUser.get().getAuthResponse().id_token;

// ... pass id token to server for validation and app sign in
```

**Comment by tiennguyen1293** [comment]
It's not working on 13.x.x.

But it's worked before Big Sur macOS release.

After i updated Big Sur and Safari 14.x.x ... it work again.

Do we have an idea to fix that?

**Comment by matheusbaumgart** [comment]
Safari 14.0.3 and still no fix. If `Prevent cross-site tracking` is enabled it doesn't work, if it **is enabled** then it works. I can't ask my users to disable it just so they can get their events from Google Calendar on my app.

**Comment by matheusbaumgart** [comment]
What am I missing? This is a massive blocker for PWAs on iOS. Even the most basic example: https://developers.google.com/calendar/quickstart/js won't work on Safari. Just open a new tab and you'll see the user is no longer authenticated.

Please shine some light, I'm lost :/

**Comment by wasyyyy** [comment]
@matheusbaumgart @tiennguyen1293 

Due to the ITP policy, only the default popup flow works on Safari. And users would appear to be signed out when they land on my pages, even they have signed in via Google before. On Chrome, they can be automatically signed in even on a different device.

Are you using the popup flow or redirect flow? I am not exactly familiar with PWAs, but assumed it should be the similar to the Safari or iOS webview. I tried the API on Safari and it works if I interact with the popup window. But if I use the redirect flow, then it still cannot give me the tokens after redirecting back to my page.

I am using Safari 14.0.2 by the way.

**Comment by veeralpatel** [comment]
Apple seems to blocking third party cookies by default now: https://webkit.org/blog/10218/full-third-party-cookie-blocking-and-more/. It seems like a big deal to get this resolved as it will affect more and more people.

**Comment by wasyyyy** [comment]
Hi, @veeralpatel 

The Sign in with Google on Safari is relying on user interacting with the Sign In popup to get the cookie access. This feature is not mentioned in recent documents, but in the original ITP 2.0 blogpost back in 2019. That being said, as long as you see the Sign In popup from Google on Safari, it should still work.

If you find things otherwise, it may indicate a bug either in the library or client implementation. It would be great to provide steps to reproduce the issue here. If that is not feasible, then could you indicate which version of Safari you are using and in what setup? Do you see any error messages printed in the developer console?

**Comment by veeralpatel** [comment]
@wasyyyy 
Yes, we see the popup come up, but after the user selects the account and the pop up closes, none of the callbacks fire. We have seen this happen the most on mobile Safari 13.1 and believe it's related to the webkit change mentioned above: https://webkit.org/blog/10218/full-third-party-cookie-blocking-and-more/.

**Comment by wasyyyy** [comment]
Hi @veeralpatel 

Just noticed that the temporary workaround is also mentioned under the Developer Guidance section from your pasted blogpost:

> Option 3: The temporary compatibility fix for popups, see section “Temporary Compatibility Fix: Automatic Storage Access for Popups” in our ITP 2.0 blog post. This compatibility fix allows the third-party to open a popup from your website and upon a tap or click in that popup gain temporary cookie access under the opener page on your website. Note that this compatibility fix will go away in a future version of Safari so only go this route if it saves you time and allows for a graceful transition period.

The Sign in with Google popup has been working under this workaround since the initial ITP 2.0 launch. If you saw the popup and selected an account, the cookies should be accessible and the issue should not be caused by the webkit change mentioned above.

Could you give more information on how you use GAPI? Do you call gapi.auth2.init() or gapi.auth2.authorize()?

One quick check we can do here is to clear all SessionStorage and LocalStorage for accounts.google.com and go through the sign in flow again. Do you see a request that goes to https://accounts.google.com/o/oauth2/iframerpc with an action=issueToken in the query?

Or do you have a link to the page where I can try out the sign in flow?

**Comment by wasyyyy** [comment]
And when you test the flow after clearing the storage, please also double check that you are not explicitly blocking 3p cookie access in Safari. This will cause the GAPI library fail to initialize and an error will be returned when you call gapi.auth2.init().

**Comment by matheusbaumgart** [comment]
@wasyyyy I could be wrong but isn't the whole point of the issue here that standard Safari users now have the "Prevent cross-site tracking" option on by default? I can't ask my users to make sure they are not blocking 3rd party cookies and uncheck a "security" feature set by default on their browser.

**Comment by veeralpatel** [comment]
@matheusbaumgart Yep, I'm on the same page as you. https://github.com/google/google-api-javascript-client/issues/589#issuecomment-820689936

**Comment by veeralpatel** [comment]
Are there any updates here? We are increasingly getting users complaining about not being able to sign in on Safari.

**Comment by wasyyyy** [comment]
@matheusbaumgart Yes you are correct that "Prevent cross-site tracking" option is now on by default for all users. Which is the purpose of ITP. But this option being on does not break Google Sign In. I tested on Safari 14.0.2 with the following steps.

1. Go to Safari privacy settings and make sure "Prevent cross-site tracking" is enabled.
2. Go to a website that has Google Sign In using Safari.
3. Click on the "Sign in with Google" button and in the popup window, choose an account (This step is crucial, there must be a popup window instead of redirection and an interaction must be made explicitly).
4. Sign in succeeds.

Could you help to test the above steps on your website and a few other websites? If you find a combination that consistently reproduces the issue, it would be very helpful for debugging.

**Comment by dvanderb** [comment]
This looks like a question about Google Login and Authentication. Please address this question to the google-signin tag on stackoverflow(https://stackoverflow.com/questions/tagged/google-signin)

If you believe this is in error, please reopen the issue.
